                                                                      ANALYSE RH
                                                    VOIR POURQUOI LES EMPLOYEES QUITTENT L'ENTREPRISE?
                                                      Dataset - WA_Fn-UseC_-HR-Employee-Attrition

Problematique: Quel profil d'employé est le plus susceptible de quitter l'entreprise — et quels leviers l'entreprise peut-elle actionner pour réduire le turnover ?

In [3]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
print("Setup complete")

Setup complete


                                                                              AUDIT

In [4]:
data = './data/WA_Fn-UseC_-HR-Employee-Attrition.csv'
data_employees = pd.read_csv(data)

In [5]:
data_employees.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


In [6]:
data_employees.info()

<class 'pandas.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   Age                       1470 non-null   int64
 1   Attrition                 1470 non-null   str  
 2   BusinessTravel            1470 non-null   str  
 3   DailyRate                 1470 non-null   int64
 4   Department                1470 non-null   str  
 5   DistanceFromHome          1470 non-null   int64
 6   Education                 1470 non-null   int64
 7   EducationField            1470 non-null   str  
 8   EmployeeCount             1470 non-null   int64
 9   EmployeeNumber            1470 non-null   int64
 10  EnvironmentSatisfaction   1470 non-null   int64
 11  Gender                    1470 non-null   str  
 12  HourlyRate                1470 non-null   int64
 13  JobInvolvement            1470 non-null   int64
 14  JobLevel                  1470 non-null   int64
 15

In [7]:
data_employees.describe()

,Age,DailyRate,DistanceFromHome,Education,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,HourlyRate,JobInvolvement,JobLevel,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
count,1470.000000,1470.000000,1470.000000,1470.000000,1470.0,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,...,1470.000000,1470.0,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000
mean,36.923810,802.485714,9.192517,2.912925,1.0,1024.865306,2.721769,65.891156,2.729932,2.063946,...,2.712245,80.0,0.793878,11.279592,2.799320,2.761224,7.008163,4.229252,2.187755,4.123129
std,9.135373,403.509100,8.106864,1.024165,0.0,602.024335,1.093082,20.329428,0.711561,1.106940,...,1.081209,0.0,0.852077,7.780782,1.289271,0.706476,6.126525,3.623137,3.222430,3.568136
min,18.000000,102.000000,1.000000,1.000000,1.0,1.000000,1.000000,30.000000,1.000000,1.000000,...,1.000000,80.0,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
25%,30.000000,465.000000,2.000000,2.000000,1.0,491.250000,2.000000,48.000000,2.000000,1.000000,...,2.000000,80.0,0.000000,6.000000,2.000000,2.000000,3.000000,2.000000,0.000000,2.000000
50%,36.000000,802.000000,7.000000,3.000000,1.0,1020.500000,3.000000,66.000000,3.000000,2.000000,...,3.000000,80.0,1.000000,10.000000,3.000000,3.000000,5.000000,3.000000,1.000000,3.000000
75%,43.000000,1157.000000,14.000000,4.000000,1.0,1555.750000,4.000000,83.750000,3.000000,3.000000,...,4.000000,80.0,1.000000,15.000000,3.000000,3.000000,9.000000,7.000000,3.000000,7.000000
max,60.000000,1499.000000,29.000000,5.000000,1.0,2068.000000,4.000000,100.000000,4.000000,5.000000,...,4.000000,80.0,3.000000,40.000000,6.000000,4.000000,40.000000,18.000000,15.000000,17.000000


In [8]:
data_employees.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='str')

In [9]:
data_employees.shape

(1470, 35)

In [5]:
data_employees.value_counts()

Age  Attrition  BusinessTravel     DailyRate  Department              DistanceFromHome  Education  EducationField  EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  JobRole                    JobSatisfaction  MaritalStatus  MonthlyIncome  MonthlyRate  NumCompaniesWorked  Over18  OverTime  PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  StandardHours  StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager
41   Yes        Travel_Rarely      1102       Sales                   1                 2          Life Sciences   1              1               2                        Female  94          3               2         Sales Executive            4                Single         5993           19479        8                   Y       Yes       11                 3                  1                         80            

                                                                                NETTOYAGE

In [7]:
# Encoder attritio en binaire
data_employees["Attrition"] = data_employees["Attrition"].map({"Yes": 1, "No": 0})
data_employees.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,1,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,0,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,1,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,0,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,0,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


In [8]:
# Vérifier les colonnes inutiles
for col in data_employees.columns:
    print(col, data_employees[col].nunique())
# si ca sort 1 donc colonnes inutiles

Age 43
Attrition 2
BusinessTravel 3
DailyRate 886
Department 3
DistanceFromHome 29
Education 5
EducationField 6
EmployeeCount 1
EmployeeNumber 1470
EnvironmentSatisfaction 4
Gender 2
HourlyRate 71
JobInvolvement 4
JobLevel 5
JobRole 9
JobSatisfaction 4
MaritalStatus 3
MonthlyIncome 1349
MonthlyRate 1427
NumCompaniesWorked 10
Over18 1
OverTime 2
PercentSalaryHike 15
PerformanceRating 2
RelationshipSatisfaction 4
StandardHours 1
StockOptionLevel 4
TotalWorkingYears 40
TrainingTimesLastYear 7
WorkLifeBalance 4
YearsAtCompany 37
YearsInCurrentRole 19
YearsSinceLastPromotion 16
YearsWithCurrManager 18


In [10]:
# nettoyer les colonnes inutiles
delete_col = data_employees.drop(['EmployeeCount', 'Over18', 'StandardHours'], axis=1, inplace=True)
print("success deleting columns")

KeyError: "['EmployeeCount', 'Over18', 'StandardHours'] not found in axis"

In [12]:
for col in data_employees.columns:
    print(col, data_employees[col].nunique())

Age 43
Attrition 2
BusinessTravel 3
DailyRate 886
Department 3
DistanceFromHome 29
Education 5
EducationField 6
EmployeeNumber 1470
EnvironmentSatisfaction 4
Gender 2
HourlyRate 71
JobInvolvement 4
JobLevel 5
JobRole 9
JobSatisfaction 4
MaritalStatus 3
MonthlyIncome 1349
MonthlyRate 1427
NumCompaniesWorked 10
OverTime 2
PercentSalaryHike 15
PerformanceRating 2
RelationshipSatisfaction 4
StockOptionLevel 4
TotalWorkingYears 40
TrainingTimesLastYear 7
WorkLifeBalance 4
YearsAtCompany 37
YearsInCurrentRole 19
YearsSinceLastPromotion 16
YearsWithCurrManager 18


In [13]:
# Verifier si les colonnes de satisfactions sont bien en int
verified = data_employees.dtypes
print(verified)

Age                         int64
Attrition                   int64
BusinessTravel                str
DailyRate                   int64
Department                    str
DistanceFromHome            int64
Education                   int64
EducationField                str
EmployeeNumber              int64
EnvironmentSatisfaction     int64
Gender                        str
HourlyRate                  int64
JobInvolvement              int64
JobLevel                    int64
JobRole                       str
JobSatisfaction             int64
MaritalStatus                 str
MonthlyIncome               int64
MonthlyRate                 int64
NumCompaniesWorked          int64
OverTime                      str
PercentSalaryHike           int64
PerformanceRating           int64
RelationshipSatisfaction    int64
StockOptionLevel            int64
TotalWorkingYears           int64
TrainingTimesLastYear       int64
WorkLifeBalance             int64
YearsAtCompany              int64
YearsInCurrent

                                                                    Analyse par agrégation

Nous allons analyser dans un premier temps les profiles des employees qui partent

In [14]:
# Quel département a le taux d'attrition le plus élevé ?
taux = data_employees.groupby('Department')['Attrition'].sum()
print(taux)

Department
Human Resources            12
Research & Development    133
Sales                      92
Name: Attrition, dtype: int64


D'apres l'analyse le departement ayant le taux d;attrition le plus eeleve est le sdepartement de Research & Development

In [15]:
# Les employés qui font des heures sup (`OverTime`) partent-ils plus ?
hours_employee = data_employees.groupby('OverTime')['Attrition'].mean()
print(hours_employee)

OverTime
No     0.104364
Yes    0.305288
Name: Attrition, dtype: float64


                                                                Interpretation du resultat

Ici on voit clairement que 30,53 des employees partent en faisant des heures supplementaires et que 10,44 des employees faisant des heures sup ne partent pas

In [17]:
# Quel est l'âge moyen des employés qui quittent vs qui restent ?
mean_age = data_employees.groupby('Attrition')['Age'].mean()
print(mean_age)

Attrition
0    37.561233
1    33.607595
Name: Age, dtype: float64


                                                                    Interpretation du resultat
Nous remarquons que l'age moyen des employees qui partent est de 33-34 et moins et ceux qui restent de 37-38 ans
En conclusion les jeunes partenet le plus

In [18]:
# Le niveau de salaire (`MonthlyIncome`) influence-t-il le départ
salary_income = data_employees.groupby('Attrition')['MonthlyIncome'].mean()
print(salary_income)

Attrition
0    6832.739659
1    4787.092827
Name: MonthlyIncome, dtype: float64


                                                                    Interpreation du resultat
Nous voyons clairement que les employees qui partent gagnent au plus 4787 de salaire a contrarion de ceux qui restent qui gagnent au moins en 6000 et plus
En conclusion ceux qui partenet gagnent moins que ceux qui restent

NOUS ALLONS MINATENANT FAIRE L'ANALYSE SELON LA SATISFACTION DES EMPLOYEES

In [20]:
# Y a-t-il un lien entre `JobSatisfaction` et `Attrition` ?
relation = data_employees[['JobSatisfaction', 'Attrition']].corr()
print(relation)

                 JobSatisfaction  Attrition
JobSatisfaction         1.000000  -0.103481
Attrition              -0.103481   1.000000


Interpretation

In [21]:
# Les employés peu satisfaits de leur équilibre vie pro/perso (`WorkLifeBalance`) partent-ils plus ?
satistfaction = data_employees.groupby('Attrition')['WorkLifeBalance'].mean()
print(satistfaction)

Attrition
0    2.781022
1    2.658228
Name: WorkLifeBalance, dtype: float64


Interpretation

In [ ]:
# Quel rôle joue `YearsAtCompany` et `YearsSinceLastPromotion`
